File: C/other/Find_the_numbers.c
//Program to find the Unique numbers in an array out of which two numbers occur exactly once and are distinct.
// You need to find the other two numbers and print them in ascending order.

#include <stdio.h> 

void findUniquePair(int arr[], int size) 
{ 
	// XOR each element and get XOR of two unique elements(ans) 
	int XOR = arr[0]; 
	for (int i = 1; i < size; i++) 
		XOR = XOR ^ arr[i]; 

	// Get a set bit of XOR (We get the rightmost set bit) 
	int set_bit = XOR & ~(XOR-1); 

	// Now divide elements in two sets by comparing rightmost 
	int x = 0, y = 0; // Initialize missing numbers 
	for (int i = 0; i < size; i++) 
	{ 
		if (arr[i] & set_bit)
			x = x ^ arr[i]; //first unique number in arr[]
		else
			y = y ^ arr[i]; //second unique number in arr[]
	} 
	if(x>y)
		printf("Distinct Numbers are %d and  %d", y, x);
	else
		printf("Distinct Numbers are %d and %d", x, y);
	
} 

int main() 
{ 
	
	int size;
	printf("Enter Size:");
	scanf("%d", &size);
	int arr[size];
	printf("Enter elements:\n");
	for(int i = 0; i < size; i++)
	{
		scanf("%d", &arr[i]);
	}
	findUniquePair(arr, size); 
	return 0; 
} 
/*
Sample I/O:
Input:
	Enter Size:6
	Enter elements:
	1 2 3 2 1 4

Output:
	Distinct Numbers are 3 and 4

Time Complexity: O(size)
Space Complexity: O(1)
*/

File: Python/other/find_the_numbers.py
# Python program to find the unique numbers in an array out
# of which two numbers occur exactly once and are distinct.
# You need to find the other two numbers and print them in ascending order.

if __name__ == '__main__':
    n = int(input("Enter the number of elements: "))
    arr = list(map(int, input().split(' ')[:n]))
    xors = 0
    for i in range(0, n):
        xors = xors ^ arr[i]
    xors = xors & ~(xors - 1)
    x = 0
    y = 0
    for i in range(0, n):
        if arr[i] & xors > 0:
            x = x ^ arr[i]
        else:
            y = y ^ arr[i]
    if x > y:
        swap(x, y)
    print("The distinct numbers are ", x, " and ", y)

# Time Complexity: O(N)
# Space Complexity: O(1)

# Input
# Enter the number of elements: 6
# 2 3 1 4 2 1

# Ouput
# The distinct numbers are  3  and  4
